### PR TITLE
A playground UI intended for the standalone controller

### PR DESCRIPTION
**Pull request by rabbah**
A UI providing playground functionality for authoring functions and running them in the browser.

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [x] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by rabbah** [comment]
Thanks @chetanmeh for adding the `--pg` option. I just tried it. ❤️

**Comment by chetanmeh** [comment]
Below is how the ui looks like

image

**Comment by chetanmeh** [comment]
Changed the behaviour to launch Playground UI by default (can be disabled with `--no-pg` flag). Also updated the readme with screenshot see https://github.com/nimbella-corp/openwhisk/tree/playground/core/standalone

PR is now ready for review

File: core/standalone/README.md [hunk 1]
```diff
@@ -102,21 +107,35 @@ $ java -jar openwhisk-standalone.jar -h
...
+                                      stores enable this flag to get bootstrap
+                                      done
       --kafka-docker-port  <arg>      Kafka port for use by docker based
                                       services. If not specified then 9091 or
                                       some random free port (if 9091 is busy)
                                       would be used
       --kafka-port  <arg>             Kafka port. If not specified then 9092 or
                                       some random free port (if 9092 is busy)
                                       would be used
+      --no-pg                         Disable Playground UI
```
[the remark this answers is not in the document]
Comment (chetanmeh) [reply core/standalone/README.md]: Makes sense. Name it `--no-ui` now

**Comment by chetanmeh** [comment]
> One small thing is, I noticed that it pulls nightly runtime images while invoking an action if I did not configure any runtime.json.
> How about pulling them at the bootup time?

Yes this aspect needs to be improved. I was thinking to explicitly pull at least the nodejs image before auto  launching the ui in browser so that by the time UI loads its immediately functional

**Comment by chetanmeh** [comment]
@style95 I have now updated the logic to explicitly prepull the nodejs default image before launching the browser. So after running `$ docker image rm openwhisk/action-nodejs-v10:nightly` the browser now gets launched only after the pre pull completes and user can start using it.

Please try again

**Comment by style95** [comment]
@chetanmeh Thank you for the update.
The reason why all runtime images are not being pulled is to speed up the bootup?

File: core/standalone/src/main/scala/org/apache/openwhisk/standalone/StandaloneDockerSupport.scala [hunk 2]
```diff
@@ -129,6 +129,18 @@ object StandaloneDockerSupport {
     else hostIpLinux
   }
 
+  def prePullImage(imageName: String)(implicit logging: Logging): Unit = {
+    //docker images openwhisk/action-nodejs-v10:nightly
+    //REPOSITORY                    TAG                 IMAGE ID            CREATED             SIZE
+    //openwhisk/action-nodejs-v10   nightly             dbb0f8e1a050        5 days ago          967MB
+    val imageResult = s"$dockerCmd images $imageName".!!
```
Comment (style95) [inline core/standalone/src/main/scala/org/apache/openwhisk/standalone/StandaloneDockerSupport.scala]: 👍

File: core/standalone/src/main/scala/org/apache/openwhisk/standalone/PlaygroundLauncher.scala [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    Await.result(f, 5.minutes)
+    Try {
+      if (!devMode) {
+        prePullDefaultImages()
+      }
+      launchBrowser(pgUrl)
+    }.failed.foreach(t => logging.warn(this, "Failed to launch browser " + t))
+  }
+
+  private def launchBrowser(url: String): Unit = {
```
Comment (style95) [inline core/standalone/src/main/scala/org/apache/openwhisk/standalone/PlaygroundLauncher.scala]: Love this thoughtful behavior. 👍

**Comment by rabbah** [comment]
Good call on using /ui. We already have this route reserved in the bundled nginx conf. Eventually could serve the playground there as well.

**Comment by chetanmeh** [comment]
> Good call on using /ui.

Change to `ui` was only done for cli options. The url is still referring to `/playground`. Should we change that also?

**Comment by tysonnorris** [comment]
url I see is `/playground/ui/index.html`

RE pulling images - when I launched standalone, the ui shows before pulling is done and I get several `Rescheduling Run message, too many message in the pool` errors when I run (and run via ui doesn't complete). I'm not sure if it is possible to not launch the ui till after the pull completes, but that may be better first time experience for button mashers. At some point it might be cool to have some status endpoint that indicates the first time setup progress (pulling images, creating prewarms, coordinate prewarm config with ui to indicate which types may be slow, etc), this would allow ui to launch, but in a appropriately disabled state till setup is complete. I wouldn't hold up this PR for any of this.

Great stuff! Thanks @rabbah  and @chetanmeh !

**Comment by chetanmeh** [comment]
> when I launched standalone, the ui shows before pulling is done

@tysonnorris did you tried by building the latest from this PR branch or via the pre built jar from [fork][1]? As [few days ago][2] I added logic to delay launching the UI till at least nodejs image pull is complete. The jars in my fork are bit old, would update them

[1]: https://github.com/chetanmeh/incubator-openwhisk/releases
[2]: https://github.com/apache/openwhisk/pull/4686#issuecomment-544366312

**Comment by tysonnorris** [comment]
@chetanmeh yes I checked out master and merged from PR branch; 
```
git checkout -b nimbella-corp-playground master
git pull https://github.com/nimbella-corp/openwhisk.git playground
```

here are the logs indicating run failures before pulling completed - I don't see the logs indicating prepull happened though ```[2019-10-23T13:34:29.255Z] [INFO] [#tid_13AlNDg3NYiYfuK55mcIiNZEFis5j7NM] [MemoryArtifactStore]  [marker:database_getDocument_finish:30048:0]
[2019-10-23T13:34:29.275Z] [ERROR] [#tid_13AlNDg3NYiYfuK55mcIiNZEFis5j7NM] [ContainerPool] Rescheduling Run message, too many message in the pool, freePoolSize: 0 containers and 0 MB, busyPoolSize: 4 containers and 1024 MB, maxContainersMemory 1024 MB, userNamespace: whisk.system, action: ExecutableWhiskAction/whisk.system/playground-run@0.0.1, needed memory: 256 MB, waiting messages: 2
[2019-10-23T13:34:33.366Z] [INFO] [#tid_2fcLKUlymbxpvX7S8koH9mAoYQDuY23A] [MemoryArtifactStore] [GET] 'activations' finding document: 'id: whisk.system/0d634632513f4c07a34632513fcc07ae' [marker:database_getDocument_start:15039]
[2019-10-23T13:34:33.366Z] [INFO] [#tid_2fcLKUlymbxpvX7S8koH9mAoYQDuY23A] [MemoryArtifactStore]  [marker:database_getDocument_finish:15039:0]
[2019-10-23T13:34:33.755Z] [INFO] [#tid_7ufom2S4DcjJ7OAuMlVnPKHRIggPCM2h] [MemoryArtifactStore] [GET] 'activations' finding document: 'id: whisk.system/4b29fddbe8704422a9fddbe870d42232' [marker:database_getDocument_start:45066]
[2019-10-23T13:34:33.755Z] [INFO] [#tid_7ufom2S4DcjJ7OAuMlVnPKHRIggPCM2h] [MemoryArtifactStore]  [marker:database_getDocument_finish:45066:0]
[2019-10-23T13:34:35.446Z] [INFO] [#tid_2bn0VghttmXsPv9xIxpWDmfEjcbkKhQ9] [WebActionsApi]  [marker:controller_blockingActivation_finish:60188:60059]
[2019-10-23T13:34:35.451Z] [INFO] [#tid_2bn0VghttmXsPv9xIxpWDmfEjcbkKhQ9] [BasicHttpService] [marker:http_post.202_counter:60192:60192]
[2019-10-23T13:34:37.509Z] [INFO] [#tid_sid_invokerWarmup] [DockerForMacClient]  [marker:invoker_docker.pull_finish:69379:66218]
[2019-10-23T13:34:37.509Z] [INFO] [#tid_sid_invokerWarmup] [DockerForMacClient]  [marker:invoker_docker.pull_finish:69380:66219]
[2019-10-23T13:34:37.510Z] [INFO] [#tid_sid_unknown] [StandaloneDockerContainerFactory] Pulled OpenWhisk provided image openwhisk/action-nodejs-v10:nightly
[2019-10-23T13:34:37.510Z] [INFO] [#tid_sid_unknown] [StandaloneDockerContainerFactory] Pulled OpenWhisk provided image openwhisk/action-nodejs-v10:nightly
[2019-10-23T13:34:37.510Z] [INFO] [#tid_sid_unknown] [StandaloneDockerContainerFactory] Pulled OpenWhisk provided image openwhisk/action-nodejs-v10:nightly
[2019-10-23T13:34:37.510Z] [INFO] [#tid_sid_unknown] [StandaloneDockerContainerFactory] Pulled OpenWhisk provided image openwhisk/action-nodejs-v10:nightly
[2019-10-23T13:34:37.510Z] [INFO] [#tid_sid_unknown] [StandaloneDockerContainerFactory] Pulled OpenWhisk provided image openwhisk/action-nodejs-v10:nightly
[2019-10-23T13:34:37.510Z] [INFO] [#tid_sid_unknown] [StandaloneDockerContainerFactory] Pulled OpenWhisk provided image openwhisk/action-nodejs-v10:nightly```

**Comment by style95** [comment]
JFYI, I could run UI after pulling the image is done.

**Comment by chetanmeh** [comment]
@tysonnorris I have improved the logic to now explicitly pull `nightly` tagged images. 

So far the logic was explicitly checking via `docker images` command to check if image exist locally. This works fine for normal tagged images which in general do not get updated. However as nightly image do get updated this logic did not worked properly. Now for nightly even if the image is present locally it would be explicitly pulled.

Please try again with latest PR branch state

**Comment by rabbah** [comment]
I will merge this once travis passes.

**Pull request merged**